### ISSUE TITLE
detect multimodal distribution in sample submission, then do not use the `mean`

### ISSUE THREAD

**Issue by jgehrcke** [conbench/conbench#535]
Two interesting cases:

![image](https://user-images.githubusercontent.com/265630/207337025-52022ef3-a37d-46ec-9acf-8b1e3b023035.png)

----------

![image](https://user-images.githubusercontent.com/265630/207337186-ae4dacd7-c0ac-463d-b71e-56c4a0adb20c.png)

---------

In both cases we got samples from three iterations.

In both cases it appears as if the mean value is either not meaningful or even misleading, or both. It appears to be largely affected by the magnitude of a warm-up effect. That is, when one plots the mean value of these scenarios over time, it seems likely that one actually mainly analyzes the severity of the warm-up effect (and not so much the 'good samples').

I think that it's obvious that given these samples the benchmark method should be investigated for methodological sanity by a human.

We could automatically detect this situation. And then do something useful. Warn users. Maybe do a fall-back by picking `min` values. Maybe emit something in the lines of a 'sample distribution alert'. 

I really like that we have examples like this in some of our environments, it shows the potential of attempting to perform per-sample distribution analysis from N iterations: https://github.com/conbench/conbench/issues/530

**Comment by jonkeane** [comment]
I would be really curious what benchmark you found this in (I don't think I see it above — but maybe I'm missing it!).

One thing we've never totally taken seriously is the warm-up effect + how to discount it (or not, depending on what you care about!). This was one of the biggest reasons [we have the disk cache clearing](https://github.com/conbench/conbench/blob/508267f8159f0b932f81b942eb5fe0e5dccf4869/conbench/runner.py#L306-L323) between each iteration on our macro benchmarks was as a way to basically remove the warmed up benefit. It was expeditious (+ has seemingly served us well), but I'm not totally sure it's the right long term solution.

As we're working on history and comparison updates — I do think we should be able to (soon) start looking at comparing differences not just in means, but also in mins like you described. I could also imagine comparing the mean of iterations 2:N, the max, and the result of just iteration 1. Each of these would have slightly different interpretations of course ("what does this look like after warm up", "what does this look like in a more real-world scenario where I run one thing once and not repeatedly", "what's the worst-case scenario" etc.). I could even imagine having toggles on the various plots + comparison pages that let you change those comparisons.

**Comment by jgehrcke** [comment]
Thanks for getting back on this!

> I would be really curious what benchmark you found this in (I don't think I see it above — but maybe I'm missing it!).

Can't answer this right now (I should have taken notes elsewhere!). All the more motivation to detect such 'bad sample' scenarios automatically :) .

I didn't include this above because I wasn't sure how public we want that to be. 

I basically brute-forced through the UI of a Conbench instance and quickly manually sanity-checked individual samples, basically trying to get a feeling for the 'quality' of raw data that we currently collect. I quickly found super interesting (problematic :)) examples and started creating tickets / collecting ideas for helping with those situations.

**Comment by jgehrcke** [comment]
> This was one of the biggest reasons [we have the disk cache clearing](https://github.com/conbench/conbench/blob/508267f8159f0b932f81b942eb5fe0e5dccf4869/conbench/runner.py#L306-L323) between each iteration on our macro benchmarks was as a way to basically remove the warmed up benefit.

Yeah, saw that! I am not sure if page cache clearing has/had an impact on the above's benchmark.

But I think in general the following is true:

If disk cache clearing has significant impact on the measured duration then this proves that disk I/O has a significant impact on the measured duration. That's a fair thing to measure _if one really wants to measure it_. However, 
- rarely that's really interesting
- comparability across benchmark runs then has stricter requirements than in cases when no or very little disk I/O is involved. E.g., becomes rather unstable when the (virtual) disk is network-attached.

> It was expeditious (+ has seemingly served us well), but I'm not totally sure it's the right long term solution.

Yeah. To be honest, I am skeptical if dropping this cache between benchmark iterations _by default_ is wise. For 'ignorant' benchmark authors, it may hide the fact that their benchmark is dominated by I/O. 

A clear warm-up effect at least allows for automatically detecting it, and after all there's higher likelihood for the measurement method / benchmark specifics to be reviewed and overhauled.

**Comment by jgehrcke** [comment]
This is such an important collection of topics :heart: 

I want to go back to what I feel like is the heart of this ticket: warning users. Do we agree that given e.g.  `[12.0, 2.0, 2.0]` the _most important thing_ to do is for a human to manually review the benchmarking method? I mean, of course we can try to be smart about this situation and work with `min` instead of `mean`. But that is just a band-aid, and may (rarely) be wrong, too.

**Comment by jonkeane** [comment]
> I want to go back to what I feel like is the heart of this ticket: warning users. Do we agree that given e.g. [12.0, 2.0, 2.0] the most important thing to do is for a human to manually review the benchmarking method? I mean, of course we can try to be smart about this situation and work with min instead of mean. But that is just a band-aid, and may (rarely) be wrong, too.

Hmm, I'm not totally sure we always want to warn in that case `[12.0, 2.0, 2.0]`. Like we discussed before: warm-up effects are common enough (though for other reasons, the way macro-benchmarks are run for arrow today with cache clearing, we generally don't observe them — or I guess really we _always_ measure them in each iteration :P ) we might want to be able to disable that _in benchmarks where we expect it_.

A slightly different case that I think could|should have a more categorical warning would be `[2.0, 12.0, 2.0]` since that isn't at all a warm-up effect, and would indicate something else is up.

So I wonder if this is something we want a setting for: allow a user (benchmark author? conbench administrator? viewer? Probably not that last one, but it's an interesting question if this is something that should be set in the benchmark running code and sent to conbench, or something set post-hoc in conbench) mark a benchmark as "this is expected to have a warmup effect, don't warn me about that, but do warn me about non-first iteration outliers", "warn me about any and all outliers", or (perhaps with strong admonition to not do this) "warn me about no outliers [I know that this benchmark is all over the place, but it's what I got]"

**Comment by austin3dickey** [comment]
I think this would make a good setting, and a good place for that setting is in https://github.com/conbench/benchalerts (which we'll monorepo into this repo in the future). That package is still pretty minimal but it's intended to hold all the logic that helps benchmark-implementers customize how and when they get alerts.

**Comment by jonkeane** [comment]
You're absolutely right @austin3dickey that if the intention of the alert is to be bubbled up to github and actively alert developers, it should go into benchalerts

This hasn't been made explicit anywhere else yet, but in this context it's become clear to me: we actually have two different places where what we're describing as an alert might go (maybe we should chose different words for these to help differentiate them): 

1. things that trigger an external notification of some sort (typically via github) to flag that something in the run of benchmarks needs to be looked at
2. things that are flagged in the UI of conbench

As it stands, we have overlap between those for many, many of these circumstances (e.g. if a benchmark is out of the z-score bounds, that shows up in the conbench UI *and* generates a github-alert for someone to look at it). But we have a handful of other circumstances that are slightly different: when making a comparison across [two benchmarks that have different contextes](https://conbench.ursa.dev/compare/benchmarks/c8a9817832d34b64a13d72b3901d2faf...d74e04fb30004001a4e72ccc8ef8a52f/) we flag in red the parts of the context that are different (if that happens across a contender-baseline for a run that benchalerts is alerting for there is (or should be) a message mentioning that the context is slightly different so it couldn't find a matching baseline, but my point is that that's slightly different from the alert we see in the conbench UI)

IMO, I suspect that we will and should move to a world where there are a number of possible alerts that happen in the UI, and then the choice as to whether to percolate them up to github via benchalerts is done with settings.  This way, one can explore the data in the conbench UI and you get the nice experience of conbench flagging down when things look funky, but we don't have a large number of alerts bubbling up through github (or at least, the level of alerts that bubble up is tuneable by implementations). As with testing and CI: we have a very fine line we have to walk where we alert when something has gone wrong, but if we alert too much people will ignore it or disable it. This alert _might_ end up being something that benchalerts should flag, but we would need *a lot* more information about how common that is (and how helpful, how noisy, etc) before we should commit to that being _the_ one place for it. 

So with all of that being said: I would advocate for us to explore what this would look like (possibly offline, possibly with an experiment in the conbench UI), see if it's helpful for developers. And if it is helpful and we gather data to see how good our true positive to false positive alerting is, we might consider adding this in to benchalerts as a bubble-up-able alert in github.

**Comment by jgehrcke** [comment]
> A slightly different case that I think could|should have a more categorical warning would be [2.0, 12.0, 2.0] since that isn't at all a warm-up effect, and would indicate something else is up.

We seem to agree on that, and that's really my main point with this ticket as of the original title: "detect multimodal distribution in sample submission, then do not use the mean". (I also agree that when one starts doing _that_ that within this approach one can find an optimization that treats the _first_ iteration's result in a special way, considering warm-up effect).

I have just found another great example for where we currently have sooo noisy data in a Conbench instance that the `mean` value becomes unrepresentative of the 'good' samples:

![image](https://user-images.githubusercontent.com/265630/212331510-3abf412b-584e-434c-bc35-9fcee73950a9.png)

It seems like this fluctuation behavior has always been rather stable because I think the z score is rather innocent. The raw data however suggests (brutal wording, yes) that benchmark method is rather useless, until the multi modal distribution is detected addressed.
